make sure only the thread that holds the GIL does Python stuff

chillenb opened the pull request:
This is one way to resolve #106.

You could also redefine
```
    check_signal_() = []() {
        PyGILState_STATE gstate;
        gstate = PyGILState_Ensure();
        if (PyErr_CheckSignals() != 0)
            throw py::error_already_set();
        PyGILState_Release(gstate);
    };
```

hczhai commented:
Thanks for identifying and solving this problem. I have the following suggestions.

1. the unnecessary ``{``, ``}``, and the comment line should be removed, so only 4 lines should be necessary.
2. In https://github.com/block-hczhai/block2-preview/commit/0708415eaa15032817177390521fb2793f9d55ef I added the ``driver.get_csf_coefficients`` in to the Python unit test. On GitHub actions, these tests run on Python 3.12, and you will see the segment fault in the log. You may merge that commit into this PR and this PR is supposed to fix the unit tests.
3. It seems that the problem was only reproducible in Python 3.12 (so it seems that they are making GIL more awkward to use in 3.12 so that they would have more reason to remove it in 3.13).

chillenb commented:
OK, I rebased off of https://github.com/block-hczhai/block2-preview/commit/0708415eaa15032817177390521fb2793f9d55ef and removed the braces/comments. Let's hope the tests pass!!

> It seems that the problem was only reproducible in Python 3.12 (so it seems that they are making GIL more awkward to use in 3.12 so that they would have more reason to remove it in 3.13).

This makes sense :laughing: It would be nice if they had gotten rid of the GIL going from Python 2 to 3--that was quite painful and GIL problems wouldn't have made it much worse.
